main.nf

```diff
@@ -3,119 +3,147 @@
+	    file("${data}/unmelted_sdrf.tsv") into UNMELTED_SDRF_QUERY
         """
         unmelt_condensed.R\
-                -i ${condensed_sdrf}\
-                -o ref_sdrf_proc.tsv\
-                --retain-types\
-                --has-ontology                 
+                -i ${data}/condensed-sdrf.tsv\
+                -o ${data}/unmelted_sdrf.tsv\
+                --has-ontology\
+                --retain-types ${params.data_import.unmelt_sdrf.retain_types}
+	# rename data dir name to avoid downstream file name collision
+	parallel cp -R ${data} ::: ${data}.query ${data}.ref 
         """
-    }
-    
+    } 
+    TRAINING_DATA_UNMELT.set{ TRAINING_DATA_PROCESSED }
 }else{
-   	REF_10X_DIR = Channel.fromPath(params.ref_10x_dir, checkIfExists: true).first()
-   	QUERY_10X_DIR = Channel.fromPath(params.query_10x_dir, checkIfExists: true).first()
-   	UNMELT_SDRF_REF = Channel.fromPath(params.unmelt_sdrf_ref, checkIfExists: true).first()
-   	UNMELTED_SDRF_QUERY = Channel.fromPath(params.unmelt_sdrf_query, checkIfExists: true).first()
-    	REF_MARKERS = Channel.fromPath(params.ref_markers, checkIfExists: true).first()
+    TRAINING_DATA.set{ TRAINING_DATA_PROCESSED }
+	}
+	
+	// fork queue channel contents into channels for corresponding tools
+	TRAINING_DATA_PROCESSED.into{
+	    GARNETT_DATA
+	    SCMAP_CELL_DATA
+	    SCMAP_CLUSTER_DATA
+	    SCPRED_DATA
+	}
+	
+	// add number of clusters to Garnett data 
+	GARNETT_FULL_DATA = GARNETT_DATA.merge(N_CLUST)
+
+}else{
+	// manual data input (Cross-Validation)
+	
+	MANUAL_INPUT_DATA = Channel.fromPath("${params.input_data}/*")
+	BARCODE_COL = Channel.from(params.metadata.query_barcode_col_name).first()
+	CELL_LABEL_COL = Channel.from(params.metadata.query_label_col_name).first()
+
+	// add dataset id and matrix type to tuple	
+	MANUAL_INPUT_DATA.map{it -> tuple(it, it.getBaseName().toString().split('\\.')[0], it.getBaseName().toString().split('\\.')[2] ) }.set{ZIP_FILES}
+	//group by matrix type
+	ZIP_FILES.groupTuple(by:[1, 2]).map{it -> tuple(it[0][0], it[0][1], it[1], it[2]) }.set{GROUPED_DATA}
+	
+	// unzip CV data
+	process unzip_data {
+	
+	input: 
+	tuple file(test_zip), file(train_zip), val(dataset_id), val(matrix_type) from GROUPED_DATA
+	val(barcode_col) from BARCODE_COL
+	val(cell_label_col) from CELL_LABEL_COL	
+	
+	output: 
+	tuple file("*.test.*"), file("*.train.*"), val(dataset_id), val(barcode_col), val(cell_label_col), val(matrix_type) into DATA
+	file("*.test.*/marker_genes_*") into MARKERS
+	file("*.test.*/unmelted_sdrf.tsv") into UNMELTED_SDRF_QUERY
+	"""
+	# unzip test 
+	test_zipdir=\$(unzip -qql $test_zip | head -n1 | tr -s ' ' | cut -d' ' -f5- | sed 's|/.*||')
+	unzip $test_zip
+	# unzip train 
+	unzip $train_zip
+	"""
+	}
+	
+	// extract marker's number of clusters value	
+	MARKERS.map{it -> it.getBaseName().toString().split('\\_')[2] }.first().set{N_CLUST}
```

Comment:
Why not use information about the number of clusters that you extract from the initial file with datasets? Actually, the chances are we won't need them after the marker files are updated.

Reply:
When passing cross-validation data from the parent workflow onto the child, CSV file does not get read, and hence cluster number has to be extracted from the parent input.zip files